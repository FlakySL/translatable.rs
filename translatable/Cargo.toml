[package]
name = "translatable"
description = "A robust internationalization solution for Rust featuring compile-time validation, ISO 639-1 compliance, and TOML-based translation management. "
repository = "https://github.com/FlakySL/translatable.rs"
license = "MIT OR Apache-2.0"
readme = "../README.md"
version = "0.1.0"
edition = "2024"
authors = ["Esteve Autet <esteve@memw.es>", "Chiko <chiko@envs.net>"]
keywords = [
    "i18n",
    "translations",
    "idioms",
    "languages",
    "internazionalization",
]

[dependencies]
strum = { version = "0.27.1", features = ["derive"] }
thiserror = "2.0.12"
translatable_proc = { path = "../translatable_proc" }
translatable_shared = { path = "../translatable_shared/" }

[dev-dependencies]
trybuild = "1.0.104"
