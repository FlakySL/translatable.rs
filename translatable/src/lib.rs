mod error;

// Export the private error module
pub use error::RuntimeError as Error;
/// Re-export the language in the crate top level as
/// it's a macro parameter
pub use shared::Language;
/// Re-export the procedural macro for crate users
pub use translatable_proc::translation;

/// This module re-exports structures used by macros
/// that should not but could be used by the users
/// of the library
pub mod shared {
    pub use strum::IntoEnumIterator;
    /// Re-export utils used for both runtime and compile-time
    pub use translatable_shared::{
        Language,
        LanguageIter,
        TranslationNode,
        TranslationNodeCollection,
        TranslationNodeError,
    };
}
