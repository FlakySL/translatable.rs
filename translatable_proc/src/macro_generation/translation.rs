use proc_macro2::TokenStream as TokenStream2;
use quote::{ToTokens, quote};
use thiserror::Error;
use translatable_shared::Language;

use crate::data::translations::load_translations;
use crate::macro_input::input_type::InputType;
use crate::macro_input::translation::TranslationMacroArgs;
use crate::utils::collections::map_to_tokens;
use crate::utils::errors::handle_macro_result;

#[derive(Error, Debug)]
enum MacroCompileError {
    #[error("The path '{0}' could not be found")]
    PathNotFound(String),

    #[error("The language '{0:?}' ('{0:#}') is not available for the path '{1}'")]
    LanguageNotAvailable(Language, String),
}

pub fn translation_macro(input: TranslationMacroArgs) -> TokenStream2 {
    let translations = handle_macro_result!(load_translations());

    if let InputType::Static(language) = input.language() {
        if let InputType::Static(path) = input.path() {
            let static_path_display = path.join("::");

            let translation_object = translations
                .find_path(&path)
                .ok_or_else(|| MacroCompileError::PathNotFound(static_path_display.clone()));

            let translation =
                handle_macro_result!(translation_object).get(language).ok_or_else(|| {
                    MacroCompileError::LanguageNotAvailable(
                        language.clone(),
                        static_path_display.clone(),
                    )
                });

            return handle_macro_result!(translation).into_token_stream();
        }
    }

    let language = match input.language() {
        InputType::Static(language) => language.clone().to_token_stream(),
        InputType::Dynamic(language) => quote! {
            translatable::shared::Language::from(#language)
        },
    };

    let translation_object = match input.path() {
        InputType::Static(path) => {
            let static_path_display = path.join("::");

            let translation_object = translations
                .find_path(path)
                .ok_or_else(|| MacroCompileError::PathNotFound(static_path_display.clone()));

            let translations_tokens = map_to_tokens(handle_macro_result!(translation_object));

            quote! {
                #[doc(hidden)]
                let path: Vec<_> = vec![#(#path.to_string()),*];

                #translations_tokens
            }
        },

        InputType::Dynamic(path) => {
            let translations_tokens = translations.to_token_stream();

            quote! {
                #[doc(hidden)]
                let path: Vec<_> = #path;

                #translations_tokens
                    .find_path(&path)
                    .ok_or_else(|| translatable::Error::PathNotFound(path.join("::")))?
            }
        },
    };

    quote! {
        (|| -> Result<String, translatable::Error> {
            std::result::Result::Ok({
                #[doc(hidden)]
                let language = #language;

                #translation_object
                    .get(&language)
                    .ok_or_else(|| translatable::Error::LanguageNotAvailable(language, path.join("::")))?
                    .to_string()
            })
        })()
    }
}
