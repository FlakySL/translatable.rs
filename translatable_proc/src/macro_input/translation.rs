use core::panic;
use std::collections::HashMap;

use proc_macro2::TokenStream as TokenStream2;
use quote::ToTokens;
use syn::parse::{Parse, ParseStream};
use syn::token::Static;
use syn::{
    Error as SynError,
    Expr,
    ExprLit,
    Ident,
    Lit,
    LitStr,
    Path,
    PathArguments,
    Result as SynResult,
    Token,
    parse2,
};
use thiserror::Error;
use translatable_shared::Language;

use super::input_type::InputType;

/// Used to represent errors on parsing a `TranslationMacroArgs`
/// using `parse_macro_input!`.
///
/// The enum implements a helper function to convert itself
/// to a `syn` spanned error, so this enum isn't directly
/// exposed as the `syn::Error` instance takes place.
#[derive(Error, Debug)]
enum TranslationMacroArgsError {
    /// An error while parsing a compile-time String value
    /// was found.
    #[error("The literal '{0}' is an invalid ISO 639-1 string, and cannot be parsed")]
    InvalidIsoLiteral(String),

    /// Extra tokens were found while parsing a static path for
    /// the `translation!` macro, specifically generic arguments.
    #[error("This translation path contains generic arguments, and cannot be parsed")]
    InvalidPathContainsGenerics,
}

/// The `TranslationMacroArgs` struct is used to represent
/// the `translation!` macro parsed arguments, it's sole
/// purpose is to be used with `parse_macro_input!` with the
/// `Parse` implementation the structure has.
pub struct TranslationMacroArgs {
    /// Represents the user specified language
    /// which may be static if the specified language
    /// is a string literal or a `Language` enum tagged
    /// union instance, otherwise dynamic and represented
    /// as a `TokenStream`.
    language: InputType<Language>,

    /// Represents a toml path to find the translation
    /// object in the previously parsed TOML from the
    /// translation files, this can be static if specified
    /// as `static path::to::translation` or dynamic if
    /// it's another expression, this way represented as a
    /// `TokenStream`.
    path: InputType<Vec<String>>,

    /// Stores the replacement arguments for the translation
    /// templates such as `Hello {name}` if found on a translation.
    ///
    /// If a call such as `a` is found, it will be implicitly
    /// converted to `a = a` thus stored like so in the hash map.
    replacements: HashMap<String, TokenStream2>,
}

impl TranslationMacroArgsError {
    pub fn into_syn_error<T: ToTokens>(self, span: T) -> SynError {
        SynError::new_spanned(span, self.to_string())
    }
}

/// The implementation is used to achieve the
/// sole purpose this structure has, which is being
/// used in a `parse_macro_input!` call.
impl Parse for TranslationMacroArgs {
    fn parse(input: ParseStream) -> SynResult<Self> {
        let parsed_langauge_arg = match input.parse::<Expr>()? {
            Expr::Lit(ExprLit { lit: Lit::Str(literal), .. }) => {
                match literal.value().parse::<Language>() {
                    Ok(language) => InputType::Static(language),

                    Err(_) => Err(TranslationMacroArgsError::InvalidIsoLiteral(literal.value())
                        .into_syn_error(literal))?,
                }
            },

            other => InputType::Dynamic(other.into_token_stream()),
        };

        input.parse::<Token![,]>()?;

        let parsed_path_arg = match input.parse::<Static>() {
            Ok(_) => {
                let language_arg = input
                    .parse::<Path>()?
                    .segments
                    .into_iter()
                    .map(|segment| match segment.arguments {
                        PathArguments::None => Ok(segment.ident.to_string()),

                        other => Err(TranslationMacroArgsError::InvalidPathContainsGenerics
                            .into_syn_error(other)),
                    })
                    .collect::<Result<Vec<_>, _>>()?;

                InputType::Static(language_arg)
            },

            Err(_) => InputType::Dynamic(input.parse::<Expr>()?.to_token_stream()),
        };

        let mut replacements = HashMap::new();
        if input.peek(Token![,]) {
            while !input.is_empty() {
                input.parse::<Token![,]>()?;
                let key = input.parse::<Ident>()?;
                let value = match input.parse::<Token![=]>() {
                    Ok(_) => input.parse::<Expr>()?.to_token_stream(),

                    Err(_) => key.clone().into_token_stream(),
                };

                replacements.insert(key.to_string(), value);
            }
        }

        Ok(Self {
            language: parsed_langauge_arg,
            path: parsed_path_arg,
            replacements,
        })
    }
}

impl TranslationMacroArgs {
    #[inline]
    #[allow(unused)]
    pub fn language(&self) -> &InputType<Language> {
        &self.language
    }

    #[inline]
    #[allow(unused)]
    pub fn path(&self) -> &InputType<Vec<String>> {
        &self.path
    }

    #[inline]
    #[allow(unused)]
    pub fn replacements(&self) -> &HashMap<String, TokenStream2> {
        &self.replacements
    }
}
