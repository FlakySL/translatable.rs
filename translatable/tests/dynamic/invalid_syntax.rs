use translatable::translation;

fn main() {
    // Missing arguments
    let lang = "es";
    let _ = translation!(lang);

    // Invalid literal type
    let lang = 42;
    let path = "common.greeting";
    let _ = translation!(lang, path);

    // Malformed dynamic path
    // TODO: I'm 100% sure that this is a bug.
    let lang = "es";
    let path = "invalid.path";
    assert!(translation!(lang, path).is_err());
}
