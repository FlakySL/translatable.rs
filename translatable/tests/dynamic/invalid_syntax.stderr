error: expected `,`
 --> tests/dynamic/invalid_syntax.rs:6:13
  |
6 |     let _ = translation!(lang);
  |             ^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `translation` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `String: From<{integer}>` is not satisfied
  --> tests/dynamic/invalid_syntax.rs:11:13
   |
11 |     let _ = translation!(lang, path);
   |             ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<{integer}>` is not implemented for `String`
   |
   = help: the following other types implement trait `From<T>`:
             `String` implements `From<&String>`
             `String` implements `From<&mut str>`
             `String` implements `From<&str>`
             `String` implements `From<Box<str>>`
             `String` implements `From<Cow<'_, str>>`
             `String` implements `From<char>`
   = note: required for `{integer}` to implement `Into<String>`
   = note: this error originates in the macro `translation` (in Nightly builds, run with -Z macro-backtrace for more info)
